### PR TITLE
Request group complex sync

### PR DESCRIPTION
**Pull request by mcanetti**
@artsalliancemedia/thunderstorm

[USERS-24](https://artsalliancemedia.atlassian.net/browse/USERS-24)

### CODE REVIEW

File: thunderstorm_auth/setup.py [hunk 1]
```diff
@@ -92,6 +93,15 @@ def _register_sync_tasks(celery_app, db_session, group_models):
         celery_app.register_task(sync_task)
 
 
+def _request_sync_task(celery_app, db_session, group_models):
```
Comment (will-norris) [inline thunderstorm_auth/setup.py]: maybe `_request_complex_groups_broadcast`

File: thunderstorm_auth/setup.py [hunk 2]
```diff
@@ -92,6 +93,15 @@ def _register_sync_tasks(celery_app, db_session, group_models):
...
+    # send a request of sync if any of group_models' table in db is empty
+    if any([db_session.query(group_model).count() == 0 for group_model in group_models]):
```
Comment (will-norris) [inline thunderstorm_auth/setup.py]: Try `db_session.query(group_model).first()` - it'll be faster than I'd say

File: thunderstorm_auth/setup.py [hunk 3]
```diff
@@ -92,6 +93,15 @@ def _register_sync_tasks(celery_app, db_session, group_models):
...
+        celery_app.send_task(
+            # TODO: @mcanetti - move this to the `ts_auth.group` exchange?
```
Comment (will-norris) [inline thunderstorm_auth/setup.py]: Using the `ts_auth.group` exchange makes sense to me I think - lets keep auth stuff living in the same place until we decide to move it all

File: test/test_setup.py [hunk 4]
```diff
@@ -65,3 +67,18 @@ def test_init_group_sync_queue(celery_app, models, group_types):
     } == {
         (setup.EXCHANGE, 'role.data'), (setup.EXCHANGE, foo_type.routing_key), (setup.EXCHANGE, bar_type.routing_key)
     }
+
+
+def test__request_sync_task(celery_app, db_session, models):
```
Comment (will-norris) [inline test/test_setup.py]: Normally I would say you should avoid testing private methods (something I have been doing lately): https://softwareengineering.stackexchange.com/questions/380287/why-is-unit-testing-private-methods-considered-as-bad-practice

File: thunderstorm_auth/setup.py (hunk 1 above)
Comment (mcanetti) [reply thunderstorm_auth/setup.py]: going for _request_complex_group_republish

File: thunderstorm_auth/setup.py (hunk 2 above)
Comment (mcanetti) [reply thunderstorm_auth/setup.py]: thanks, the first() it's a better choice. Following conversation on routing key, I'm not sure we want to differentiate at this point. The future complex-group.data will probably have both complex members and user members so make sense that a complex_group_user_association model would also trigger a complex-group.republish/complex-group.data message exchange

File: thunderstorm_auth/setup.py (hunk 3 above)
Comment (mcanetti) [reply thunderstorm_auth/setup.py]: leaving ts-messaging as it is the right exchange. '{}.ts_auth.group'.format(celery_app.main) is queue created by the consumers.

File: test/test_setup.py (hunk 4 above)
Comment (mcanetti) [reply test/test_setup.py]: normally I would think the same, but we do the same for the other private methods like _init_group_sync_tasks and _init_group_sync_queue. I only just remembered to add the second underscore in the test method.

**Pull request merged**